README.md
# alfr

[![Travis build status](https://travis-ci.com/rwetherall/alfr.svg?branch=master)](https://travis-ci.com/rwetherall/alfr)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/rwetherall/alfr?branch=master&svg=true)](https://ci.appveyor.com/project/rwetherall/alfr)

## Overview

'alfr' allows you to connect to an 'Alfresco' content management repository and interact with its contents using simple and intuative functions.

## Installation

You can install the development version from github with:

```r
# install.packages("remotes")
remotes::install_github("rwetherall/alfr")
```


NEWS.md
## alfr 1.2.0

* Added folder and document delete
* Added session invalidate
* Added session is valid check
* Use pkgdown to generate package site documentation
* Improved examples
* Unit test refactor
   - tests atomic
   - mocked API easy to refresh
   - coverage > 90%

## alfr 1.1.0

* Improved documentation

## alfr 1.0.0

* Establish connection to Alfresco server
* Node retrieval and creation.
* Content upload.
* Content download.


tests/testthat/helper.R
library(httptest)

.mockPaths("../")

set_requester(function (request) {
  gsub_request(request, "http://localhost:8080/alfresco/api/-default-/public/", "api/")
})

test_execution_mode <- "mock"

test_server <- "http://localhost:8080"

admin_username <- "admin"
admin_password <- "admin"

mocked_test_that <- function (desc, code) do_mocking(test_that(desc, code), test_execution_mode)

do_mocking <- function (expr, mode=c("mock", "live", "capture"))
  match.arg(mode) %>%
  switch(
    mock = with_mock_api(expr),
    live = eval(expr),
    capture = capture_requests(expr))


tests/testthat/testalf_node.R
context("alf-node")

require(magrittr)
require(httptest)

test_doc_path <- "Sites/test/documentLibrary/testFolder/testdoc.txt"
test_doc_id <- "b569f89f-e5a0-4ef4-8c4e-d1314bac67a4"

# TODO Given an invalid path to a node, When I try to retieve the document by path, Then I am unsucessful

# TODO get folder

mocked_test_that(

  "Given a valid path to a document,
   When I try to retieve the documeny by path,
   Then I am successful and can access information about that node", {

  node <- alf_session(test_server, admin_username, admin_password) %>% alf_node(relative_path = test_doc_path)

  expect_false(is.null(node$id))
  expect_equal(node$name, "testdoc.txt")

})

mocked_test_that(

  "Given a document with plain text content set,
   When ask for the content as a file,
   Then the file is created and I can read its content", {

  node <- alf_session(test_server, admin_username, admin_password) %>% alf_node(relative_path = test_doc_path)
  content <- node$content

  expect_false(is.null(content))

  expect_equal(content$mime_type, "text/plain")
  expect_equal(content$mime_type_name, "Plain Text")
  expect_equal(content$size, 19)
  expect_equal(content$encoding, "ISO-8859-1")

  if (test_execution_mode == "live") {
    content_file <- content$as.file() %>% file("r")
    expect_false(is.null(content_file))
    readLines(content_file) %>% startsWith("Test txt document.") %>% expect_true()
    close(content_file)
  }
})

# TODO get cbinary content (eg word, pdf)
# TODO get image
# TODO get binary content that can be converted to data (eg CSV, excel)

mocked_test_that(

  "Given a document with plain text content set,
   When I try to upload content from a local file,
   Then the content is updated and the node node details are available", {

  node <- alf_session(test_server, admin_username, admin_password) %>%
          alf_node(relative_path = test_doc_path)
  updated_node <- node$content$update("resources/testuploaddoc.txt")

  expect_false(is.null(updated_node))

  expect_equal(updated_node$content$mime_type, "text/plain")
  expect_equal(updated_node$content$mime_type_name, "Plain Text")
  expect_equal(updated_node$content$size, 41)
  expect_equal(updated_node$content$encoding, "ISO-8859-1")

  if (test_execution_mode == "live") {
    content_file <- updated_node$content$as.file() %>% file("r")
    expect_false(is.null(content_file))
    readLines(content_file) %>% startsWith("Somthing") %>% expect_true()
    close(content_file)
  }
})



tests/testthat/testalf_session.R
context("alf-session")

require(magrittr)
require(httptest)

mocked_test_that(

  "Given a valid respository,
   And valid user credentials,
   When I request a session,
   Then I recieve a valid session", {

  validSession <- function (session)
    session %>% is.null() %>% not() &&
    session$ticket %>% startsWith("TICKET_") &&
    session$server == test_server

  alf_session(test_server, admin_username, admin_password) %>% validSession() %>% expect_true()
})

mocked_test_that(

  "Given a valid respository,
   And invalid user credentials,
   When I request a session,
   Then I recieve an error", {

  expect_error(alf_session(test_server, "wrong", "wrong"), "Authentication")
})

mocked_test_that(

  "Given a valid session,
   When I check it is valid,
   Then I am told that it is", {

  session <- alf_session(test_server, admin_username, admin_password)
  expect_true(alf_session.is_valid(session))
})

mocked_test_that(

  "Given an invalid session,
   When I check it is valid,
   Then I am told that it is", {

  expect_false(alf_session.is_valid(alf_sample.invalid_session))
})

mocked_test_that(

  "Given a valid session,
   When I invalidate the session,
   Then it is no longer valid", {

  session <- alf_session(test_server, admin_username, admin_password)
  result <- alf_session.invalidate(session)
  expect_true(result)
})

mocked_test_that(

  "Given an invalid session,
   When I invalidate the session,
   Then nothing happens", {

  result <- alf_session.invalidate(alf_sample.invalid_session)
  expect_false(result)
})


R/alf_http.R
require(jsonlite)
require(magrittr)

##
#' @title Alfresco GET method
#' @description Helper to make a GET call to the Alfresco REST API
#' @param endpoint base endpoint URI
#' @param ticket authentication ticket
#' @param params optional list of parameters
#' @param as return as \code{json} (default), \code{raw}
#' @param body request body
#' @return result based on format provided
#' @export
##
alf_GET <- function (endpoint, ticket, params=list(), as=c("json", "raw"), body=NULL)
  alf_method("GET", endpoint, ticket, params, as, body)

##
#' @title Alfresco POST method
#' @description Helper to make a POST call to the Alfresco REST API
#' @param endpoint base endpoint URI
#' @param ticket authentication ticket
#' @param params optional list of parameters
#' @param as return as \code{json} (default), \code{raw}
#' @param body request body
#' @return result based on format provided
#' @export
##
alf_POST <- function (endpoint, ticket, params=list(), as=c("json", "raw"), body=NULL)
  alf_method("POST", endpoint, ticket, params, as, body)

##
#' @title Alfresco PUT method
#' @description Helper to make a PUT call to the Alfresco REST API
#' @param endpoint base endpoint URI
#' @param ticket authentication ticket
#' @param params optional list of parameters
#' @param as return as \code{json} (default), \code{raw}
#' @param body request body
#' @return result based on format provided
#' @export
##
alf_PUT <- function (endpoint, ticket, params=list(), as=c("json", "raw"), body=NULL)
  alf_method("PUT", endpoint, ticket, params, as, body)

##
# @title Alfresco HTTP method
# @description Helper to make a http method call to the Alfresco REST API
# @param method HTTP method to call
# @param endpoint base endpoint URI
# @param ticket authentication ticket
# @param params optional list of parameters
# @param as return as \code{json} (default), \code{raw}
# @param body request body
# @return result based on format provided
##
alf_method <- function (method=c("GET", "POST", "PUT"), endpoint, ticket, params=list(), as=c("json", "raw"), body=NULL) {

  # check we have a valid method
  method <- match.arg(method)

  # check we have a valid as
  as <- match.arg(as)

  # construct method call
  method_call <- bquote(.(as.symbol(method))(

      # resolve parameters
      add_params(endpoint, params),

      # add authentication
      add_headers(Authorization = paste("Basic", base64_enc(ticket))),

      # add body
      body=body))

  # get response
  response <- eval(method_call)

  # check for error
  if (http_error(response))
    if(response$status_code == 401)
      stop("Authentication details are not longer valid or have timed out.  Please renew your session. (401)")
    else if (response$status_code == 409)
      stop("A duplicate node was found.  Please ensure name is unique. (409)")
    else
      stop(http_status(response)$message)

  # process response as
  switch (
    as,
    json = content(response, "text") %>% fromJSON(flatten = TRUE),
    raw =  content(response, "raw"))
}

##
# Helper method to add parameters onto a URL endpoint
##
add_params <- function (endpoint, params, sep="?") {

  if (length(params) > 0)
    add_params(
      paste(endpoint, sep, names(params[1]), "=", params[[1]], sep=""),
      params[-1], sep="&")
  else
    endpoint
}



R/alf_session.R
require(httr)
require(magrittr)

# base endpoint helper
base_endpoint <- function(server, endpoint) paste(server, "/alfresco/api/-default-/public/", endpoint, sep="")

# tickets endpoint function
tickets_endpoint <- function(server) base_endpoint(server, "authentication/versions/1/tickets")

##
#' @title
#' Get connection session to Alfresco content repository
#' @description
#' Validates authentication details with Alfresco content repository, returning ticket, server details and endpoints if
#' successful.
#' @param server Alfresco server URL
#' @param username user name
#' @param password password
#' @return Connection session to Alfresco repository
#' @examples
#' \donttest{
#' # establish a connection to the alfresco content repository
#' my_session <- alf_session("alfresco.my-org.com", "admin", "admin")
#'
#' # output the server URL
#' print(my_session$server)
#' #[1] "alfresco.my-org.com"
#'
#' # output the connection ticket
#' print(my_session$ticket)
#' #[1] "xxx-xxx-xxx-xxx"
#' }
#' @export
##
alf_session <- function (server, username, password) {

  # try to get the authentication ticket for the repository
  response <-
    tickets_endpoint(server) %>%
    POST(body=list(userId = username, password = password), encode = "json")

  # check for error
  if (http_error(response))

    # indicate authentication failed
    if (response$status_code == 403) stop("Authentication failed, please check your username and password are correct.")

    # otherwise stop with error message
    else stop(http_status(response)$message)

  # get the ticket from the response
  else {

    # node end points
    node_endpoint <- function(node_id) base_endpoint(server, "alfresco/versions/1/nodes/") %>% paste(node_id, sep="")
    node_content_endpoint <- function(node_id) node_endpoint(node_id) %>% paste("/content", sep="")
    node_children_endpoint <- function(node_id) node_endpoint(node_id) %>% paste("/children", sep="")

    # session details
    list(
      server = server,
      ticket = fromJSON(content(response, "text"), flatten = TRUE)$entry$id,
      node_endpoint = node_endpoint,
      node_content_endpoint = node_content_endpoint,
      node_children_endpoint = node_children_endpoint
    )
  }
}

# TODO
# alf_session.is_valid <- function (session) {
#
# }

#TODO
# alf_session.invalidate <- function (session) {
#
# }


R/alf_node.R
require(httr)
require(magrittr)

##
#' @title Get Alfresco node
#' @description
#' Gets the details of an Alfresco repository node matching \code{node_id} or, if provided, the node at \code{relative_path}
#' relative to \code{node_id}.
#' @param session valid Alfresco repository session
#' @param node_id node id, defaults to \code{-root-}
#' @param relative_path relative path from \code{node_id} to required node, defaults to \code{NULL}
#' @return Node details
#' @examples
#' \donttest{
#' # establish a connection to the alfresco content repository
#' my_session <- alf_session("alfresco.my-org.com", "admin", "admin")
#'
#' # get details of document node
#' my_document <- alf_node(my_session, relative_path = "testFolder/testdoc.txt")
#'
#' # output the name of the document
#' print(my_document$name)
#' #[1] "testdoc.txt"
#'
#' # output the details of documents content
#' print(my_document$content$mime_type)
#' #[1] "text/plain"
#' print(my_document$content$mime_type_name)
#' #[1] "Plain Text"
#' print(my_document$content$size)
#' #[1] 19
#' print(my_document$content$encoding)
#' #[1] "ISO-8859-1"
#'
#' # read document content
#' my_content_file <- file(my_document$content$as.file(), "r")
#' my_content <- readLines(my_content_file)
#' close(my_content_file)
#'
#' # upload new content
#' my_document <- my_document$content$update("resources/testuploaddoc.txt")
#' }
#' @export
##
alf_node <- function (session, node_id = "-root-", relative_path = NULL) {

  # TODO deal with invalid session

  # TODO deal with invalid path

  # build parameter list
  params <- list()
  if (!is.null(relative_path)) params$relativePath = relative_path

  # GET node details found at the path provided
  alf_GET(session$node_endpoint(node_id), session$ticket, params) %>%
    as.node(session)
}

##
#' @title Create a new Alfresco node
#' @description
#' Creates a new Alfresco repository node as a child of \code{node_id}.
#' @param session valid Alfresco repository session
#' @param node_id node id
#' @param node_details details of new node
#' @return node details
#' @export
##
alf_node.new <- function(session, node_id, node_details) {

  # TODO deal with invaild session

  # POST node update
  alf_POST(session$node_children_endpoint(node_id), session$ticket, body = toJSON(node_details, auto_unbox = TRUE)) %>%
    as.node(session)

}

##
# Helper to represent response as node information
##
as.node <- function (response, session) {

  # TODO determine whether this is a folder or not

  # node details
  list (
    id = response$entry$id,
    name = response$entry$name,
    content = as.content(response, session))
}

##
# Helper to represent response as content information
##
as.content <- function (response, session) {

  endpoint <- session$node_content_endpoint(response$entry$id)
  ticket <- session$ticket

  ##
  # Get content as file
  ##
  as.file <- function(destination_file = NULL) {

    # create temp file if none provided
    if (is.null(destination_file)) destination_file <- tempfile()

    # open file for binary write
    bfile <- base::file(destination_file, "wb")

    # GET content and write to file
    alf_GET(endpoint, ticket, as="raw")  %>% writeBin(bfile)

    # close and return
    close(bfile)
    destination_file
  }

  ##
  # Update content
  ##
  update <- function(file_path) {

    # check that the path provided references a file that exists
    if (!file.exists(file_path)) stop(paste("Can not upload content from", file_path, "as it does not exist."))

    # PUT the content
    response <- alf_PUT(endpoint, ticket, as="json", body=upload_file(file_path))

    # return updated node
    as.node(response, session)
  }

  list (
    mime_type = response$entry$content$mimeType,
    mime_type_name = response$entry$content$mimeTypeName,
    size = response$entry$content$sizeInBytes,
    encoding = response$entry$content$encoding,
    as.file = as.file,
    update = update
  )
}


